src/modules/ecdsa_s2c/main_impl.h

```diff
@@ -0,0 +1,199 @@
+    sha->s[6] = 0x37ee0874ul;
+    sha->s[7] = 0x421fc55ful;
+
+    sha->bytes = 64;
+}
+
+int secp256k1_ecdsa_s2c_sign(const secp256k1_context* ctx, secp256k1_ecdsa_signature* signature, secp256k1_ecdsa_s2c_opening* s2c_opening, const unsigned char
+ *msg32, const unsigned char *seckey, const unsigned char* s2c_data32) {
+    secp256k1_scalar r, s;
+    int ret;
+    unsigned char ndata[32];
+    const unsigned char* noncedata = NULL;
+    secp256k1_sha256 s2c_sha;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(signature != NULL);
+    ARG_CHECK(seckey != NULL);
+    ARG_CHECK(s2c_data32 != NULL);
+
+    /* Provide `s2c_data32` to the nonce function as additional data to
+     * derive the nonce It is first hashed because it should be possible
+     * to derive nonces even if only a SHA256 commitment to the data is
+     * known.  This is important in the ECDSA anti-klepto protocol. */
+    secp256k1_s2c_ecdsa_data_sha256_tagged(&s2c_sha);
+    secp256k1_sha256_write(&s2c_sha, s2c_data32, 32);
+    secp256k1_sha256_finalize(&s2c_sha, ndata);
+    noncedata = ndata;
+
+    secp256k1_s2c_ecdsa_point_sha256_tagged(&s2c_sha);
+    ret = secp256k1_ecdsa_sign_inner(ctx, &r, &s, NULL, &s2c_sha, s2c_opening, s2c_data32, msg32, seckey, NULL, noncedata);
+    secp256k1_scalar_cmov(&r, &secp256k1_scalar_zero, !ret);
+    secp256k1_scalar_cmov(&s, &secp256k1_scalar_zero, !ret);
+    secp256k1_ecdsa_signature_save(signature, &r, &s);
+    return ret;
+}
+
+int secp256k1_ecdsa_s2c_verify_commit(const secp256k1_context* ctx, const secp256k1_ecdsa_signature* sig, const unsigned char* data32, const secp256k1_ecdsa_s2c_opening* opening) {
+    secp256k1_ge commitment_ge;
+    secp256k1_ge original_pubnonce_ge;
+    unsigned char x_bytes[32];
+    secp256k1_scalar sigr, sigs, x_scalar;
+    secp256k1_sha256 s2c_sha;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_context_is_built(&ctx->ecmult_ctx));
+    ARG_CHECK(sig != NULL);
+    ARG_CHECK(data32 != NULL);
+    ARG_CHECK(opening != NULL);
+
+    if (!secp256k1_ecdsa_s2c_opening_load(ctx, &original_pubnonce_ge, opening)) {
+        return 0;
+    }
+    secp256k1_s2c_ecdsa_point_sha256_tagged(&s2c_sha);
+    if (!secp256k1_ec_commit(&ctx->ecmult_ctx, &commitment_ge, &original_pubnonce_ge, &s2c_sha, data32, 32)) {
+        return 0;
+    }
+
+    /* Check that sig_r == commitment_x (mod n)
+     * sig_r is the x coordinate of R represented by a scalar.
+     * commitment_x is the x coordinate of the commitment (field element).
+     *
+     * It is sufficient to only compare the x coordinates as it is as difficult to find an client
+     * commitment to the negation of the point as to any other point. There is a small reduction in
+     * security as it is easier to find a collision with a point and its negation.
+     */
+    secp256k1_ecdsa_signature_load(ctx, &sigr, &sigs, sig);
```

Comment:
What is the additional code? I need to verify the ECDSA siganture to get the full R point don't I?

Reply:
Oh ok, I understand now. I was on the wrong here track, and I believe the comment led me to onto this track.

> It is sufficient to only compare the x coordinates as it is as difficult to find an client commitment to the negation of the point as to any other point. There is a small reduction in security as it is easier to find a collision with a point and its negation.

The commitment comprises only the x-coordinate, so you can of course only verify the x-coordinate. And I would not call this a reduction in security. If you don't include y-bit, you can't expect that it provides any security.

Having said this, it's true that if the signature is valid, it implies a particular y-bit, so for *valid* signatures, you could compare the entire point. But since we separated verification of the commitment and the signature, this is not relevant for us.   

Now when I read the commitment again, it makes sense in the context of the antiklepto protocol. (Note also the term "client commitment".) But this should not bother us either since we also separated s2c and antiklepto.